add docs on using Azure cni v1 using single NIC per node (#3627)
kubernetes-sigs/cluster-api-provider-azure: docs/book/src/topics/azure-cni-v1.md

> # Using Azure CNI V1

Reviewer: Want to move this to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/topics/addons.md#cni instead of its own page?

Author: Thanks for pointing it out. This doc should be part of addons. Moving it.

This document provides step to use Azure CNI as the CNI solution in your workload cluster.

## Azure Container Networking Interface v1

While following the [quick start steps in Cluster API book](https://cluster-api.sigs.k8s.io/user/quick-start.html#quick-start), Azure CNI can be used in place of Calico as a container networking interface solution for your workload cluster.

Artifacts required for Azure CNI:

- [azure-cni.yaml](https://raw.githubusercontent.com/Azure/azure-container-networking/v1.5.3/hack/manifests/cni-installer-v1.yaml)

> <!-- TODO: will we be exposing azure cni v1 customized cluster manifest for the users to use? If no, I will put in a example manifest here for the users. -->

Reviewer: is this still TODO?

Author: Good catch! I have addressed this, and isn't a TODO. https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/addons/azure-cni-v1/kustomization.yaml should fetch azure-cni-v1 installer from the azure container networking repo every time users uses `make generate`

## Update Cluster Configuration

The following resources need to be updated if working off of `capi-quickstart.yaml` (the default flavored cluster manifest generated while following the Cluster API quick start).

- `kind: AzureCluster`
  - update `spec.networkSpecs.subnets` with the name and role of the subnets you want to use in your workload cluster.

  - ```yaml
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AzureCluster
    metadata:
      name: ${CLUSTER_NAME}
      namespace: default
    spec:
      .
      .
      networkSpec:
        subnets:
        - name: control-plane-subnet # update this as per your nomenclature
          role: control-plane
        - name: node-subnet # update this as per your nomenclature
          role: node
      .
      .
    ```

- `kind: KubeadmControlPlane` of control plane nodes

>   - add `max-pods: "110"` to `spec.kubeadmConfigSpec.initConfiguration.nodeRegistration.kubeletExtraArgs`.

Reviewer: I wonder if we should make the example use `30` max pods instead of 110 to follow the AKS default https://learn.microsoft.com/en-us/azure/aks/configure-azure-cni#maximum-pods-per-node

Reviewer: (and if so we should also update the privateIPConfigs to 30)

Author: I agree to both, updating the private IP limit to default of `30`

  - add `max-pods: "110"` to `spec.kubeadmConfigSpec.joinConfiguration.nodeRegistration.kubeletExtraArgs`.

  - ```yaml
    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    kind: KubeadmControlPlane
    metadata:
      name: ${CLUSTER_NAME}-control-plane
      namespace: default
    spec:
      kubeadmConfigSpec:
        .
        .
        initConfiguration:
          nodeRegistration:
            kubeletExtraArgs:
              max-pods: "110"
              .
              .
        joinConfiguration:
          nodeRegistration:
            kubeletExtraArgs:
              max-pods: "110"
              .
              .
    ```

- `kind: AzureMachineTemplate` of control-plane
  - Add `networkInterfaces` to controlplane's `AzureMachineTemplate`

  - ```yaml
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AzureMachineTemplate
    metadata:
      name: ${CLUSTER_NAME}-control-plane
      namespace: default
    spec:
      template:
        spec:
          .
          .
          networkInterfaces:

>           - privateIPConfigs: 110 # maximum private IPs to be allocated in this node

Reviewer: ```suggestion
          - privateIPConfigs: 110 # private IPs to be allocated for this network interface
```

    kind: AzureMachineTemplate
    metadata:
      name: ${CLUSTER_NAME}-md-0
      namespace: default
    spec:
      template:
        spec:
          networkInterfaces:
          - privateIPConfigs: 110
            subnetName: node-subnet
          .
          .
    ```

- `kind: KubeadmControlPlane` of worker nodes
  - add `max-pods: "110"` to `spec.template.spec.joinConfiguration.nodeRegistration.kubeletExtraArgs`.

  - ```yaml
    apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
    kind: KubeadmConfigTemplate
    metadata:
      name: ${CLUSTER_NAME}-md-0
      namespace: default
    spec:
      template:
        spec:
          .
          .
          joinConfiguration:
            nodeRegistration:
              kubeletExtraArgs:
                max-pods: "110"
                .
                .
    ```

## Limitations

- We can only configure one subnet per control-plane node. Refer [CAPZ#3506](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3506)

- We can only configure one Network Interface per worker node. Refer[Azure-container-networking#3611](https://github.com/Azure/azure-container-networking/issues/1945)

> <!-- TODO: Do we specify the number of IPs per nodes depending on the VM size? because Refer https://learn.microsoft.com/en-us/azure/aks/configure-azure-cni#maximum-pods-per-node -->

Reviewer: are these still TODO?

Author: Not really. Updated the docs and removed these TODOs.
